### PR TITLE
Move to Zustand and add Filters to web_test

### PR DESCRIPTION
**Pull request by SkillGG**
- Moved from Redux to Zustand
- Add Filters to Popular novels accordion
- Add rule for consistent type definitions
- Add TriState Switches and Checkboxes (used in XCheckboxFilter)

### CODE REVIEW

File: src/prettify.tsx [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export type Prettify<T> = T extends object
+  ? {
+      [K in keyof T]: T[K];
+    } & {
+      /** */
+    }
+  : T;
```
Comment (nyagami) [inline src/prettify.tsx:7]: what does this file do?

File: src/store/pluginStore.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import { Plugin } from '@typings/plugin';
+import { StoreCreator } from '.';
+import { getPlugin } from '@provider/plugins';
+
+export type PluginStore = {
+  selectedPlugin?: Plugin.PluginItem;
```
Comment (nyagami) [inline src/store/pluginStore.ts:6]:
```suggestion
  pluginItem?: Plugin.PluginItem;
```

File: src/libs/filterInputs.ts [hunk 3]
```diff
@@ -11,59 +17,84 @@ export enum FilterTypes {
...
+type ExcludableCheckboxFilter = {
   type: FilterTypes.ExcludableCheckboxGroup;
   options: readonly FilterOption[];
-  value: { include?: string[]; exclude?: string[] };
-}
+  /** Default value */
+  value: {
+    /** Checkboxes marked as included */
+    include?: string[];
+    /** Checkboxes marked as included */
```
Comment (nyagami) [inline src/libs/filterInputs.ts:52]:
```suggestion
    /** Checkboxes marked as excluded */
```

File: src/libs/filterInputs.ts [hunk 4]
```diff
@@ -11,59 +17,84 @@ export enum FilterTypes {
...
+  ? CheckboxFilter
+  : Type extends FilterTypes.Picker
+    ? PickerFilter
+    : Type extends FilterTypes.Switch
+      ? SwitchFilter
+      : Type extends FilterTypes.TextInput
+        ? TextFilter
+        : Type extends FilterTypes.ExcludableCheckboxGroup
+          ? ExcludableCheckboxFilter
+          : never);
```
Comment (nyagami) [inline src/libs/filterInputs.ts:78]: image

Can we make a map type?
Like this
```ts
type FilterMapping = {
  [FilterTypes.CheckboxGroup]: CheckboxFilter
}
export type Filter<Type extends FilterTypes> = {
  label: string;
} & FilterMapping[Type]
```

File: src/prettify.tsx (hunk 1 above)
Comment (SkillGG) [reply src/prettify.tsx:7]: https://www.totaltypescript.com/concepts/the-prettify-helper

I thought it'd work on those big types, but it doesn't, unfortunately. I forgot to delete it.

File: src/libs/filterInputs.ts (hunk 4 above)
Comment (SkillGG) [reply src/libs/filterInputs.ts:78]: I think it should work, but I'd need to test it.

File: src/store/overlayStore.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      ...state,
+      loading: true,
+    }));
+  },
+  hideLoading() {
+    set(state => ({
+      ...state,
+      loading: false,
+    }));
+  },
```
[the remark this answers is not in the document]
Comment (SkillGG) [reply src/store/overlayStore.ts:29]: zustand is global store tho. `useAppStore(state=>state.loading)`.

did you mean local component state?

File: src/store/pluginStore.ts (hunk 2 above)
Comment (SkillGG) [reply src/store/pluginStore.ts:6]: You have to change it in other files that use this car as well.

File: src/store/overlayStore.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (nyagami) [reply src/store/overlayStore.ts:29]: I meant global loading. But seems it is not very important. We can remove it xD

File: src/libs/filterInputs.ts (hunk 4 above)
Comment (nyagami) [reply src/libs/filterInputs.ts:78]: ye, just more readable code, nested code always scares people

**Pull request merged**